ako-operator/controllers/akoconfig_controller.go

```diff
@@ -0,0 +1,240 @@
+
+func (r *AKOConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("ako-operator", req.NamespacedName)
+
+	var ako akov1alpha1.AKOConfig
+	err := r.Client.Get(ctx, req.NamespacedName, &ako)
+	if err != nil && errors.IsNotFound(err) {
+		// akoconfig object got deleted, before we come here, so just return
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Info("unable to fetch AKOConfig object", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	if !ako.ObjectMeta.DeletionTimestamp.IsZero() {
+		if finalizerInList(ako.ObjectMeta.Finalizers, CleanupFinalizer) {
+			if err := r.CleanupArtifacts(ctx, log); err != nil {
+				return ctrl.Result{}, err
+			}
+			ako.ObjectMeta.Finalizers = removeFinalizer(ako.ObjectMeta.Finalizers, CleanupFinalizer)
+			if err := r.Update(ctx, &ako); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// return from here, no more reconciliation as the AKOConfig is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	// reconcile all objects
+	err = r.ReconcileAllArtifacts(ctx, ako, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AKOConfigReconciler) ReconcileAllArtifacts(ctx context.Context, ako akov1alpha1.AKOConfig, log logr.Logger) error {
+	// reconcile all the required artifacts for AKO
+	err := r.ReconcileConfigmap(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	err = r.ReconcileServiceAccount(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	err = r.ReconcileClusterRole(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	err = r.ReconcileClusterroleBinding(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	err = r.ReconcilePodSecurityPolicy(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	err = r.ReconcileStatefulSet(ctx, ako, log)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *AKOConfigReconciler) CleanupArtifacts(ctx context.Context, log logr.Logger) error {
+	objList := getObjectList()
+	for objName, obj := range objList {
+		if err := r.deleteIfExists(ctx, objName, obj); err != nil {
+			log.Error(err, "error while deleting object")
+			return err
+		}
+	}
+	objectList = nil
+	return nil
+}
+
+func (r *AKOConfigReconciler) deleteIfExists(ctx context.Context, objNsName types.NamespacedName, object runtime.Object) error {
+	err := r.Client.Get(ctx, objNsName, object)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if err == nil {
+		if err := r.Client.Delete(ctx, object); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AKOConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&akov1alpha1.AKOConfig{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&appsv1.StatefulSet{}).
+		Complete(r)
+}
+
+func (r *AKOConfigReconciler) ReconcileConfigmap(ctx context.Context, ako akov1alpha1.AKOConfig, log logr.Logger) error {
```

Comment:
Instead of using this function, can we call createOrUpdateConfigMap directly ? ReconcileConfigmap doesn't seem to be doing anything else. 
Similar for other functions bellow.

Reply:
Sure, yeah. Makes sense. Will do that.